feature: Support weighted ExtCosts

akashin opened the pull request:
This PR addresses https://github.com/near/nearcore/issues/8264, in particular allowing to add parameter weights for Storage Costs.

Some open questions and potential improvements:
- [ ] Do we need to surface weights in the `views.rs` (RPC interface of the node?)
- [ ] Ideally we would avoid specifying the old base value in the diff config if we only want to change the weight. For now we need to do:
```yaml
wasm_storage_read_base: {
    old: { base: 50_000_000_000, weight: 3 },
    new: { base: 50_000_000_000, weight: 7 }
}
```

jakmeier commented on core/primitives-core/src/config.rs:
```diff
@@ -262,9 +263,15 @@ pub struct ViewConfig {
 }
 
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct WeightedGas {
```
This naming makes it easy to be confused with https://docs.rs/near-sdk/latest/near_sdk/struct.GasWeight.html

Also, we already have "used", "prepaid", "burnt", "burned", and "attached" gas as terms used throughout the code and it often makes technical communication difficult. E.g. the question "how much gas does this TX use" already has at least 3 correct answers, depending on how you define "used gas" and transaction. Weighted gas adds one more interpretations of "used gas". :see_no_evil: 

I don't know a better name, it's probably fine to keep it. But when writing the NEP (which we should start soon) we will need to make a final decision if we have an alternative naming or if we want to keep this.

Previously brainstormed ideas for what we can call weighted gas include:
- fumes or emissions (parameter weight defines how much emissions burning gas causes)
- cycles  / compute time / chunk space (naming it after the condition checked when we include receipts in a chunk)

Again, keeping the name is probably fine. But I wanted to write this down here to give you the full context and maybe you can see a way to minimize how complicated the story around gas is.

akashin replied:
As discussed on the NEP, renamed the whole struct into
```rust
struct ParameterCost {
  gas: Gas,
  compute: u64,
}
```

I also plan to introduce a dedicated type for compute cost.

jakmeier commented on core/primitives/src/runtime/parameter_table.rs:
```diff
@@ -48,6 +59,16 @@ impl ParameterValue {
         }
     }
+
+    fn as_weighted_gas(&self) -> Option<WeightedGas> {
+        match self {
+            ParameterValue::WeightedGas { base, weight } => {
+                Some(WeightedGas { base: *base, weight: weight.into() })
+            }
+            ParameterValue::U64(v) => Some(WeightedGas { base: *v, weight: 1.into() }),
+            _ => None,
+        }
+    }
```
Ah nice, that's an elegant way of handling the old and the new format!

jakmeier commented on core/primitives/src/views.rs:
```diff
@@ -2624,7 +2624,8 @@ impl From<ExtCostsConfigView> for near_primitives_core::config::ExtCostsConfig {
                 ExtCosts::alt_bn128_pairing_check_element => view.alt_bn128_pairing_check_element,
-        };
+        }
+        .map(|_, value| WeightedGas { base: value, weight: 1.into() });
```
This might be fine for now but it makes the conversion from config to view back to config lossy. We should check if we have some tests that would fail because of that. If so, we need to figure out how to deal with that test. (I don't want tests to randomly start failing because we set the first weight != 1)

akashin replied:
I'll investigate such failures now, but I think an important question here is why do we even have a conversion back from the ConfigView to the Config? Is it used anywhere in a meaningful way? I'll try to disable it and see what breaks.

jakmeier replied:
1. So far, everything in the configuration was interesting enough to users that we wanted it to be reflected in the view and it's easy to test that by converting it back. (This argument _might_ no longer holds with the addition of param weights).
2. For some reasons I am not familiar, the Rosetta RPC adapter constructs a `RuntimeConfig` from an RPC call. There we actually need this conversion in production code.(This seems worth fixing but I don't know how much effort it would be.)

akashin replied:
I've added non-default compute cost to one of the parameters and verified that no tests in nearcore fail. I've also reviewed the code usage of `RuntimeConfigView`[1] and haven't found any examples where we do a conversion back from the view into config. I'll try removing the `From` in that direction and see if anything breaks and also will ask on Zulip in case anyone knows why it exists.

[1] https://cs.github.com/?scope=org%3Anear&scopeName=near&q=%22RuntimeConfigView%22

jakmeier replied:
Hm, maybe the usage in rosetta has been removed. Even better :D

If you can remove it and all tests pass, go for it. I don't think we have to check further if it works in the latest code base. Otherwise, if for some reason we really need the conversion, I'd vote to remove the `From` impl anyway and replace it with a constructor that clearly indicates in the name that it is a lossy conversion.

jakmeier reviewed (commented):
Looks pretty good. Having a design for how to specify these brings us a step closer to writing a NEP with reference implementation.

I am not 100% yet if we want to merge it to master already. Ideally, we don't include changes before the NEP is approved. But if it makes your work easier, then we can also merge it already as it doesn't have any effect until we actually set some weights.

> Do we need to surface weights in the views.rs (RPC interface of the node?)

I think we don't want to leak that info, it should not be of any concern to users. (If we need a way to check current weights of a running node, I think the debug page or even Grafana metrics would be more appropriate. But probably overkill and too early to add such a thing anyway.)

> Ideally we would avoid specifying the old base value in the diff config if we only want to change the weight.

Hm, I see. The current solution also works okay enough I think. Including the gas value for full context does not seem all that bad anyway.

Also, there are some CI failures in existing tests.

akashin commented:
@jakmeier , I've updated this PR according to the NEP and split it out into multiple parts.
This first part adds definition of Compute Costs to the parameter config. Given that this specific PR should not affect the behavior of nodes, do we still need to gate it behind a feature flag?
